src/main/c/NativePtr.cpp

```diff
@@ -0,0 +1,366 @@
+#include "hail/NativePtr.h"
+#include "hail/NativeObj.h"
+#include <jni.h>
+#include <assert.h>
+#include <stdio.h>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
+
+namespace hail {
+
+// JNI interface functions corresponding to Scala NativeBase and NativePtr
+
+#define NATIVEFUNC(typ) \
+  extern "C" JNIEXPORT typ JNICALL
+
+namespace {
+
+class AlignedBuf {
+public:
+  char buf_[2*sizeof(long)];
+  long force_align_;
+
+public:
+  AlignedBuf() {
+    set_addrA(0);
+    set_addrB(0);
+  }
+
+  inline NativeObjPtr& as_NativeObjPtr() {
```

Comment:
I like to use "inline" explicitly for methods which are simple enough that it's always going
to be more efficient to have them inline rather doing a call/return.  YMMV.

Reply:
My thinking was that seeing the function definition inside the class definition was explicit enough that the method is inline, but I don't feel strongly one way or the other.

Reply:
Another C++ dummy here.

Why is `inline` redundant? There's a rather [famous SO post](https://stackoverflow.com/questions/1759300/when-should-i-write-the-keyword-inline-for-a-function-method) which is the source of all my C++ `inline` knowledge. It suggests that inline is basically a signal to the linker to make sure that every translation unit uses the same function even though it might get defined in multiple translation units.

Reply:
In this case, I'm not saying anything about the semantics of `inline`, just the fact that
```C++
class Foo {
public:
  void bar() {
    // ...
  }
  // ...
};
```
is equivalent to
```C++
class Foo {
  // ...
};

inline void Foo::bar() {
  // ...
}
```
i.e. functions defined inside the class definition have an implicit `inline` annotation.